### DISCUSSION TITLE
External Tables 🙏🙏🙏

### DISCUSSION THREAD

**Discussion by jakthom** [duckdb/duckdb#14422] [Ideas]
### Overview

DataFusion has [external tables](https://datafusion.apache.org/user-guide/sql/ddl.html#create-external-table), by which paths (either local fs or remote object store) can be registered as a local table and made queryable as such. They are quite useful.

Snowflake has had [external tables](https://docs.snowflake.com/en/sql-reference/sql/create-external-table) for a long time. They are also quite useful.

GlareDB goes a step further and allows [external tables to be created across multiple data sources](https://docs.glaredb.com/reference/sql-commands/create-external-table.html), including bigquery, mongodb, postgres, snowflake, etc. It is pretty neat.

It would be very powerful to support similar patterns with DuckDB. Shipping around DuckDB database files with predefined external tables (but not the underlying data until the table is refreshed) would unlock significant amounts of additional potential.

### Why is this useful?

Creating a bunch of (empty) external tables in a named duckdb database file before shipping the db file around to clients would be a very ergonomic way to query object storage.

It would:

* Separate table representations from the underlying data and configuration. Ideally client processes shouldn't have to care about globbed parquet paths, how `some=hive/thing=is/partitioned=by/year=2024`, or how a mysterious incantation of csv reader `dateformat` parameters managed to get them `timestamp` types. They should be able to query things that look like tables (backed by what? who cares! 😁) and get data back.
* Allow locations to be modified once, centrally, without needing to change client code. A new DuckDB database file could be created and shipped to clients.
* Centralize all DDL to where it can be tested/vetted. Right now all DDL gets shipped to the client and as DuckDB's utility expands, the table definition/database "assembly" process gets pretty bloated. It would be awesome to abstract that process from clients and keep the complexity behind the scenes, which could be easily done by shipping pre-built database files.

Right now clients have to assemble and populate their own tables (from paths, databases, http endpoints, etc). It would be great to be able to assemble database files ahead of time, distribute them (without data), and have clients only need to `select` and `refresh` these predefined tables.

For example, pre-building a `duckd.db` database file from the assembly side:

```
CREATE EXTERNAL TABLE APP_LOGS
LOCATION = "s3://someapp/logs/*"
PARTITION_TYPE = HIVE
FILE_FORMAT = PARQUET
REFRESH = ON_FIRST_SELECT

CREATE EXTERNAL TABLE USERS
LOCATION = "https://somewhere.net/users"
REFRESH = ON_CREATE;

CREATE EXTERNAL TABLE ACCOUNTS
LOCATION = "s3://someapp/accounts/accounts.parquet"
REFRESH = ON_SELECT;

CREATE EXTERNAL TABLE BAD_IPS
LOCATION = "gcs://someapp/bad_ips"
FORMAT = DELTA
REFRESH 10s;

```

And on the consumption side, immediately after firing up the preexisting `duckd.db` file:

```
$ duckdb duckd.db


D desc tables;
┌──────────┐
│   name   │
│ varchar  │
├──────────┤
│ accounts │
│ app_logs │
│ bad_ips  │
│ users    │
└──────────┘


-- Explicitly refresh tables
alter table accounts refresh;
alter table app_logs refresh;
alter table bad_ips refresh;
alter table users refresh;

-- Query them immediately (without needing to build the tables, know underlying source or object storage paths, etc)
select
  *
from
  app_logs al
join
  users u
on
  al.userId = u.id
left join
  bad_ips bi
on
  al.ipAddress = bi.ip
....
```

### Further Ramblings/Opinions

* I really don't think the number of supported sources would have to be particularly large to have a big impact. Local file, s3, and gcs-backed external tables would probably suffice for a very long time.
* If only parquet-backed external tables were supported the feature would still be fantastically powerful.
* Being able to specify partitioning scheme, even if it was [only hive](https://duckdb.org/docs/data/partitioning/hive_partitioning.html), would be super useful.
* Having the option to specify Delta/Iceberg when creating the external table (to switch things to `delta_scan` and `iceberg_scan`, respectively) would be nice.
* Having the option to `refresh on creation`, `refresh on first select`, or manually (like Snowflake does with `alter table $TBL refresh;`) would be neat.
* This feature would not require `write`-side optimization to be useful. The feature could be `read`-side only and would still be potent. Being able to flush external tables, even without compaction, would also be useful.

**Comment by Maxxen** [comment]
Can you not use a [VIEW](https://duckdb.org/docs/sql/statements/create_view.html) for this? The only downside is that you can't specify paths that are relative  _from the db file_, but presumably you're not going to ship the database file _and_ a bunch of data files in a predetermined directory layout anyway. If all you're doing is referencing data on cloud storage, a view should suffice. You might find [this article](https://www.nikolasgoebel.com/2024/05/28/duckdb-doesnt-need-data) interesting.

**Reply by jakthom** [reply]
@Maxxen awesome post, thanks for sharing.

To answer your q -> yes-ish. Views can definitely handle much of this use case but it would be great to go slightly further. 

1. I want to be able to more accurately control the client-side caching of said views. The [materialized view](https://github.com/duckdb/duckdb/discussions/3638) discussion outlines this pretty well. Views query the underlying data each time, and I would very-much like to be able to control the latency hit of that. I'd rather "cache on $condition" or have explicit "refresh view" mechanisms, similar to what @xnge described.

2. Similar to ^, it would be really nice to be able to explicitly refresh/pre-cache paths. IE: "cache the past day of a hive-partitioned, s3-backed table immediately after startup using explicit `REFRESH`, but go-get-and-cache from s3 for other/more-historical partitions, depending on the query." Snowflake's external tables allow for this and it's 🔥 .

3. Datafusion's [external tables are writeable](https://datafusion.apache.org/user-guide/sql/write_options.html#write-options), which is a pretty useful thing. Sure, it doesn't do a great job with compaction/etc. But simply inserting into an s3-based "table" and getting a parquet file out the other end is a super powerful thing. I often find myself wishing for this functionality in DuckDB. Being able to do it without the client specifying any DDL would be even better. IE -> `insert into $EXTERNAL_TBL from select * from ('some_file.parquet')`

**Reply by jakthom** [reply]
Admittedly I didn't want to venture too far into a `materialized View` discussion due to the complexity of `view maintenance`.

IMO something similar to Snowflake's `refresh` functionality for `external tables` would unlock a lot of things.

[tl;dr](https://docs.snowflake.com/en/sql-reference/sql/alter-external-table) -> 

```
ALTER EXTERNAL TABLE [ IF EXISTS ] <name> REFRESH [ '<relative-path>' ]

ALTER EXTERNAL TABLE [ IF EXISTS ] <name> ADD FILES ( '<path>/[<filename>]' [ , '<path>/[<filename>'] ] )

ALTER EXTERNAL TABLE [ IF EXISTS ] <name> REMOVE FILES ( '<path>/[<filename>]' [ , '<path>/[<filename>]' ] )

ALTER EXTERNAL TABLE [ IF EXISTS ] <name> SET AUTO_REFRESH = { TRUE | FALSE }

ALTER EXTERNAL TABLE <name> [ IF EXISTS ] ADD PARTITION ( <part_col_name> = '<string>' [ , <part_col_name> = '<string>' ] ) LOCATION '<path>'

ALTER EXTERNAL TABLE <name> [ IF EXISTS ] DROP PARTITION LOCATION '<path>'
```

**Reply by Maxxen** [reply]
I think it's useful to distinguish materialized views from _incrementally updated_ views, and this seems very much like materialized views to me. I guess there is nothing in particular stopping us from implementing something like this but I think caching semantics in general is pretty hard to communicate effectively to users.

Actually thinking about it some more also raises some questions. How would we handle queries on stale views, e.g. when the underlying schema of the file changes since last refresh? Actually, does the external table schema change when refreshing as well? If so, what happens if there exists a view referencing the schema of the external table and it all of a sudden changes after a refresh so that a previously referenced column no longer exists? If we try to refresh on schema mismatch/error, do we need a stop condition so we don't endlessly poll e.g. every globbed file in the whole bucket if files keep changing between the type resolution and execution phase of the query execution?

Dealing with transactions, or rather lack of transactions when writing also seems a bit messy. Do we even support writing to an external table in an explicit transaction? What happens if we rollback? In my mind sql INSERT always carry some notion of transactionality on its own but if the duckdb host process would crash/run out of memory/exit halfway through a external table write there's not really any option of "recovering" the external table at all and any read would most likely fail due to corrupted files.

**Reply by soerenwolfers** [reply]
maybe a read only version of external tables could be offered first, with undefined behavior if the backing data is changed under duckdb's feet? that would already be supremely useful to me (being unfamiliar with the external table concept, I've previously asked for something similar at https://github.com/duckdb/duckdb/discussions/14037)

**Reply by jakthom** [reply]
@soerenwolfers 💯 agree. Really all I'm looking for is ability to cache and manage the `REFRESH` without (explicitly) re-running ddl each time. At least at first (though the mind wanders quickly).

**Reply by dforsber** [reply]
I support the idea and I think it should progressed further some way.

I jokingly suggested to use DuckDB as the data catalog, as it can create views and even define which Parquet files, for example, belong to the view (think of how Iceberg works). However, these views get pretty fast old if they are transferred to the client side in a db file, whilst on the e.g. S3 side, things progress all the time. Like, do you give a prefix for year=2024 or for all years, or should we invent something like "relative time". Adding/removing partitions also happens on the data producer side, not consumer who opens the db file.

DuckDB community has not taken very clear stance on what to do with data catalogs or how to use them. Sure, there are delta and iceberg extensions.

There is also the http proxy thingy, which would be nice to see through as a fully supported feature. DuckDB has also blob cache, but not sure if its really 1st level citizen atm.

One of the reasons, these things progress slowly is that there is no DuckDB-to-DuckDB interface - meaning like figuring out metadata "of the other end", compressed binary transport, streaming. Ability to attach remote DuckDB databases is one step closer to this, though.

It's pretty clear that these things will be implemented in way or the other. Like there is the http server extension being developed for making every duckdb instance a http server :). 

In the end you have NATs and what-not in-between, so you need to have support for control plane on higher level, like signalling etc. mediators.

Just random thoughts, probably showing that I don't have a clear suggestion forward. And thus, I think, a higher level plan is needed and then very clear and crisp implementation steps. Like for example, addressing the issues highlighted above and in other comments.

**Comment by samansmink** [comment]
I think this is a really interesting discussion, i've thought about similar things before when working on the experimental [Unity Catalog extension](https://github.com/duckdb/uc_catalog). Right now the proof-of-concept extension supports reading from delta tables. However to be able to support reading and writing to/from delta tables **and** iceberg tables cleanly, we would need to make changes to how catalogs currently work internally.

My initial idea was to implement a sort-of `MetaCatalog`, which can hold child Catalogs, which are then responsible for holding one or more tables of the types they understand. We could consider leveraging this mechanism to also implement external tables in the regular DuckDB catalog. Perhaps we could consider wrapping the regular DuckDB catalog in a MetaCatalog where we also allow adding more catalogs for external tables in parquet, iceberg, delta or whatever format. 

DuckDB is already [pretty strong](https://duckdb.org/2024/01/26/multi-database-support-in-duckdb.html) in multi-database support, ideally we could expand that with not only strong support to external catalogs like Unity Catalog or maybe Polaris, while also providing its own flexible catalog that can be stored in DuckDB database files.

**Reply by jakthom** [reply]
Totally agree w/r/t cataloging - it's one of my favorite use cases at the moment. Being able to pre-build catalogs in the form of `duckd.db` files (really databases without data), ship them to dark corners, and have client code be as dumb as possible w/r/t what actually backs the catalog is 🔥.

I could be wrong, but I suspect a `MetaCatalog` might not be necessary for basic `external table` functionality.

IE:

`create external table <T> as <DDL>` -> is effectively a `create view <V> as <DDL>` statement
`refresh external table <T>` -> is effectively a `create table <T> as select * from <V>` statement

The (basic) external table workflow is very-much doable in DuckDB now. But in order to populuate `<T>`  clients have to explicitly understand :
* the underlying view `<DDL>` (source, syntax, integration, secret, etc)
* to create a `view` first, and that `views` will never cache
* to create a `table` using `CTAS` from ^ view

It would be very ergonomic to have a (client-side) workflow that looks more like:

1. `refresh external table <T>`
2. `select * from <T>`

**Comment by teaguesterling** [comment]
Something else that could possibly be supported by external tables that isn't addressed by views would be defining foreign keys and other sorts of constraints on the data. While that obviously can't be enforced on the external data, it would be quite useful for tools that would introspect the DuckDB schema. Additionally, it may be a rather natural way to do data validation of a schema against external tables.

**Comment by ccollingwoodAutodesk** [comment]
Using duckdb as a "catalog" has been one of our primary implementations (in addition to the compute engine obviously). Our first implementation was to initialize the "catalog" on start-up (serverless, statelesss) before handing the connection over to the service. The initialization consisted of preparing extensions, credentials, running some DDLs, virtual views (registering Python relations/arrow-scanners), all of which is pretty quick and easy. However, it is still overhead and requires a certain amount of domain-specific knowledge and understanding. 

Most of this could be operationalized into a prepared/maintained/evolving `.db` file (along with a `hint` file for quickly identifying the most recent version) stored on S3 and simply attached at start-up. But setting up the virtual views (relations/scanners/datasets) or external tables (iceberg metadata file) is still a post-attach responsibility. 

I could imagine ways that this wouldn't be an issue for iceberg (and probably some other ext table formats). Unfortunately for iceberg, there isn't a `refresh`-like option since a lot of iceberg management systems seem to be moving away from producing `hint.txt` files that would greatly simplify such a responsibility (perhaps intentional and by design in case there is a need/requirement to time-travel). Atm, providing the relevant `.metadata.json` file is still required. This file could of course could be passed as an env-variable. Or the refresh-step would simply involve scanning/listing the metadata directory seaching for the most recent `.metadata.json` file (last_modified). Neither of which is very ergonomic if that is one of the objectives of such an effort.

The Python-specific parts (relations/dataframes/arrow-scanners) obviously introduce a run-time dependency potentially making a prepared `.db` file problematic for other runtimes (CLI, Java, R, etc) and less portable in general. 

I can again image ways to overcome this through synergy with some other projects like data-fusion/flight/adbc. Similar to how mysql/postgres/sqlite connections can be attached, additional/new attach semantics (extensions) could emerge. Perhaps this meta-catalog would provide the abstraction for such things so that specific implementations wouldn't have to be created for each and instead rely on background bridging.

Moreover, it would be incredible if this meta-catalog was not just responsible for providing ways to describe catalogs/schemas/structures, but also a means by which push-downs, logical/physical plans can be exchanged/expressed (similar discussion in python-iceberg https://github.com/apache/iceberg-python/issues/30) across query-engines/planners (similar to substrait?). Eg: projections/push-downs/etc. 

For example, such functionality might ease the way of defining an external iceberg table ahead of time. Part of the of the attach/refresh process might be simplified and handled ahead of time (instead of scan-time) improving ergonomics and runtime performance. Based on the particular iceberg metadata, the inventory of data-files could be classified (deletes, appends, mixed), compiled and cached (file pointers, not the data in the files), inspect data-layout (field-ids column names, partitions, types), etc... so that by the time the table was ready to query, most of the most the scanner has to do today is already done.

(I understand that the iceberg-ext is external to duckdb, but really pointing out that there could be a convention for external-table/catalog extensions to follow)

```sql
CREATE EXTERNAL TABLE MY_SCHEMA.MY_IB -- implicitly read-only without col/type definition
    TYPE = 'ICEBERG'
    METADATA = 's3://some_bucket/my_ib_table'  -- automatically inspects/searches for latest /metadata/???.metadata.json file
    ICEBERG_VERSION = 2
    ...
```

This might be quite similar to how a data-fusion table might be registered as an external table. The catalog would understand which columns could be projected, filters pushed-down, partitions/fragments, etc... Maybe even Arrow Flight and ADBC too.

```sql
ATTACH 'host=grpc://localhost port=8080' AS flightdb (TYPE FLIGHT);

CREATE EXTERNAL TABLE MY_SCHEMA.SOME_VW
    TYPE = 'FLIGHT' 
    SOURCE = 'flightdb.foobar';
```

Perhaps I'm creeping beyond the scope of the discussion at hand, but probably worth keeping in mind of where things may begin to drift in the future.

**Comment by mullinsms** [comment]
I like the proposal for external tables since they're available in every analytic data platform from Netezza to BigQuery. One thing I haven't seen mentioned here is secrets management for all of these external connections. In order for the DuckDB db to be shippable the connection secrets need to be encrypted in a fashion that can't be read by the user. Currently DuckDB persisted secrets are stored unencrypted on disk.

**Reply by dforsber** [reply]
> I like the proposal for external tables since they're available in every analytic data platform from Netezza to BigQuery. One thing I haven't seen mentioned here is secrets management for all of these external connections. In order for the DuckDB db to be shippable the connection secrets need to be encrypted in a fashion that can't be read by the user. Currently DuckDB persisted secrets are stored unencrypted on disk.

If you find a way todo that offline with Open Source DuckDB, you'll be rich ;). Just thinking about the dilemma that you need the decryption key in the binary to read the encrypted secrets. Wouldn't help much even if the source was closed. Its a big rabbit hole and leads to TPMs and what not. The best that I can think of is that you get tokens bound to user identities and varying scopes and what not to limit the exposure in case the token is exposed/leaked. However, that's something like token vending that should be closer to realtime than packaging. Just babbling here :).

**Reply by samansmink** [reply]
Secret storage in duckdb has been designed with extensibility of the secret storage in mind. We currently only implement 1 transient in-memory secret storage and one unencrypted file-backed one, but more can be added and there is already syntax for that (see https://github.com/duckdb/duckdb/blob/main/test/sql/secrets/create_secret_storage_backends.test):
```
CREATE SECRET my_secret IN some_secret_storage ( ... )
```
Note that Motherduck uses this to allow you to persist secrets remotely already, see https://motherduck.com/docs/sql-reference/motherduck-sql-reference/create-secret

I can definitely see us expanding DuckDB's native secret storage capabilities at some point to allow secure sharing of secrets

**Comment by cmettler** [comment]
Explicitly defined external tables could be used for additional statistics, similar to this approach 
https://learn.microsoft.com/en-us/sql/t-sql/statements/create-statistics-transact-sql?view=sql-server-ver16
Maybe also ART indexes could be created on top of external tables, if it turns out useful for some queries.

**Comment by yan-hic** [comment]
discussion has stalled unfortunately as ducklake could have supported this nicely. In our case, we would love to use ducklake to add/update on top of existing hive-partitioned parquet files (in gcs). One can create views on these parquet files, but you can't insert into them.

Ducklake should allow to _construct_ metadata based on **existing** parquet files, then you could update/add (which would create new parquet files in another location). This could be the `CREATE EXTERNAL TABLE`...

**Reply by SebZbp** [reply]
Would this apply as well to this use case?
- lots of hive partitioned parquet files with very small size in S3 (generated by a stream process); difficult and inefficient to query with a view due to large number of files
- adding those files to ducklake as an external table in read only would allow to trigger table maintenance operations (including compaction in another bucket where new repartitioned files would be stored)
- old poorly partionned data would ultimately get orphaned and could be cleaned up 
  - by an external process having write/delete access on those files
  - based on data from the ducklake catalog (source data is replaced by optimised parquet files in 2nd location)
- in parallel, data gets queryable through duckdb/ducklake and optimised storage
Currently, this kind of operation needs to be designed and executed manually.

**Comment by samansmink** [comment]
> we would love to use ducklake to add/update on top of existing hive-partitioned parquet files

@yan-hic Almost there :) https://github.com/duckdb/ducklake/pull/252


edit: ah nvm I see you already commented there

**Comment by yan-hic** [comment]
> @yan-hic Almost there :) https://github.com/duckdb/ducklake/pull/252

Almost indeed if PR included table creation if not exist.
